code/eyetracking/eyetrackingrun.py

```diff
@@ -0,0 +1,746 @@
+# Copyright 2023 The Axon Lab <theaxonlab@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+from pathlib import Path
+import string
+from typing import Optional, List, Tuple, Union
+import numpy as np
+import pandas as pd
+from scipy.signal import fftconvolve
+from pyedfread import edf, edfread
+import re
+import os
+import json
+import seaborn as sns
+import matplotlib.pyplot as plt
+from collections import defaultdict
+
+
+class EyeTrackingRun:
+    """
+
+    """
+
+    EYE_CODE_MAP = defaultdict(
+        lambda: "unknown", {"R": "right", "L": "left", "RL": "both"}
+    )
+    RIGHT_EYE_COLUMNS = {
+        "time": "eye_timestamp",
+        "gx_right": "eye1_x_coordinate",
+        "gy_right": "eye1_y_coordinate",
+        "pa_right": "eye1_pupil_size",
+        "px_right": "eye1_pupil_x_coordinate",
+        "py_right": "eye1_pupil_y_coordinate",
+        "hx_right": "eye1_head_x_coordinate",
+        "hy_right": "eye1_head_y_coordinate",
+        "rx": "screen_pixel_per_degree_x",
+        "ry": "screen_pixel_per_degree_y",
+        "gxvel_right": "eye1_x_velocity",
+        "gyvel_right": "eye1_x_velocity",
+        "hxvel_right": "eye1_head_x_velocity",
+        "hyvel_right": "eye1_head_y_velocity",
+        "rxvel_right": "eye1_raw_x_velocity",
+        "ryvel_right": "eye1_raw_y_velocity",
+        "fgxvel": "eye1_fast_gaze_x_velocity",
+        "fgyvel": "eye1_fast_gaze_y_velocity",
+        "fhxyvel": "eye1_fast_head_x_velocity",
+        "fhyvel": "eye1_fast_head_y_velocity",
+        "frxyvel": "eye1_fast_raw_x_velocity",
+        "fryvel": "eye1_fast_raw_y_velocity",
+    }
+
+    def __init__(
+        self,
+        session: int,
+        task_name: str,
+        participant: int,
+        samples: pd.DataFrame,
+        events: pd.DataFrame,
+        messages: pd.DataFrame,
+        message_first_trigger: str,
+        screen_resolution: Tuple[int, int],
+        messages_start_fixation: str = "",
+        messages_stop_fixation: str = "",
+        pe: str = "",
+    ):
+        """
+
```

Comment:
Please write a docstring... Apologies I didn't have time to write it for you and just suggested placeholders.